Add `--trace-dispatch`

kpamnany opened the pull request:
Similar to `--trace-compile`, emit the `precompile` statement for a method once, but only when it is dynamically dispatched.

For this, we add a `dispatched` field into `jl_method_instance_t` that is initialized to 0, and when the method is dispatched, it is set to 1 and the precompile statement is emitted (once).

TODO:
- [X] Add NEWS entry.
- [X] Update manual.
- [X] Add test.

kpamnany committed:
Add `--trace-dispatch`

Similar to `--trace-compile`, emit the `precompile` statement for a method
once, but only when it is dynamically dispatched.

For this, we add a `dispatched` field into `jl_method_instance_t` that is
initialized to 0, and when the method is dispatched, it is set to 1 and the
precompile statement is emitted (once).

kpamnany committed:
Add NEWS entry

kpamnany committed:
Add documentation

vtjnash commented on src/method.c:
```diff
@@ -630,6 +630,7 @@ JL_DLLEXPORT jl_method_instance_t *jl_new_method_instance_uninit(void)
     jl_atomic_store_relaxed(&mi->cache, NULL);
     mi->cache_with_orig = 0;
     jl_atomic_store_relaxed(&mi->precompiled, 0);
+    mi->dispatched = 0;
```
I think you want to reset this in staticdata too. Also should probably be made marked atomic

kpamnany replied:
As suggested by @timholy, I renamed `mi->precompiled` to `mi->flags`, made bit 0 mean `precompiled` and bit 1 mean `dispatched`.

So now `flags` are reset in staticdata, and are atomic.

vtjnash commented on src/gf.c:
```diff
@@ -2561,6 +2561,40 @@ static void record_precompile_statement(jl_method_instance_t *mi, double compila
+        }
+        else {
+            if (ios_file(&f_dispatch, t, 1, 1, 1, 1) == NULL)
+                jl_errorf("cannot open dispatch statement file \"%s\" for writing", t);
```
Very awkward (and slightly unsound) to throw errors from here. We should try to move this into the jloptions file so we can detect any issues earlier

kpamnany replied:
I essentially duplicated `record_precompile_statement` from right above this, which does the same thing.

vtjnash reviewed (commented):
I think you also need to cover the invoke codepath in this same file to get all dispatches, but looks like a good start overall to implementing this

timholy commented on src/jltypes.c:
```diff
@@ -3610,21 +3610,23 @@ void jl_init_types(void) JL_GC_DISABLED
                             "cache_with_orig",
-                            "precompiled"),
-                        jl_svec(7,
+                            "precompiled",
```
I don't think we ever use the upper bits in `precompiled`, so alternatively we could repurpose it into a bitfield. Might merit a name change, though, and perhaps that's not worth it.

kpamnany replied:
Made this change, please see if it looks okay.

timholy reviewed (commented):
Clearly needed in conjunction with #55047, thanks for doing this.

kpamnany committed:
Address review comments

- Rename `precompiled` field in `jl_method_instance_t` to `flags`
  and use bit 0 as `precompiled` and bit 1 as `dispatched`.
- Use relaxed atomic operations to access `flags`.
- Add setting of `dispatched` to `jl_gf_invoke_by_method`.

kpamnany commented:
> I think you also need to cover the invoke codepath in this same file to get all dispatches, but looks like a good start overall to implementing this

I added [this](https://github.com/JuliaLang/julia/pull/55848/files#diff-9a718f382a240acdbe9d0bdbc3ffb521f67af0e62b2ba8c8594b8faa65c78014R3512-R3518); does that look right?

Is there anywhere else where this should be recorded?

kpamnany committed:
Fix test

kpamnany committed:
Add test for `--trace-dispatch`

Also add a test for `--trace-compile`.

kpamnany commented:
Added a test.

JeffBezanson commented on src/gf.c:
```diff
@@ -3360,6 +3396,13 @@ JL_DLLEXPORT jl_value_t *jl_apply_generic(jl_value_t *F, jl_value_t **args, uint
                                                      jl_int32hash_fast(jl_return_address()),
                                                      world);
     JL_GC_PROMISE_ROOTED(mfunc);
+    uint8_t miflags = jl_atomic_load_relaxed(&mfunc->flags);
```
Are we sure this has no overhead? Could it go in a slow path?

kpamnany replied:
Good point, not sure. I think the cost should be negligible. Thoughts @vtjnash?

Could add an outer check for `jloptions.trace_dispatch`? Remove the check from inside the `record... ` function?

kpamnany replied:
I've moved this atomic load off the fast path.

JeffBezanson replied:
This is probably better but I think it should go in a slower lookup path in jl_lookup_generic; at least only when the fast cache misses.

kpamnany replied:
Ah, I see. Going by the comments in `jl_lookup_generic_`, there is a lookup in an associative cache, and if that misses, a lookup in the full cache. Can a method instance be in either of those caches if it hasn't been dispatched? If not, then I could move the call [here](https://github.com/JuliaLang/julia/blob/master/src/gf.c#L3337)?

kpamnany replied:
I've moved it [here](https://github.com/JuliaLang/julia/pull/55848/files#diff-9a718f382a240acdbe9d0bdbc3ffb521f67af0e62b2ba8c8594b8faa65c78014R3381-R3390); please take a look?

kpamnany committed:
Move the atomic load off the fast path

kpamnany committed:
Move `record_dispatch_statement` call to method cache miss path

kpamnany commented:
The test failures seem unrelated:
```
Error in testset Pkg:
Test Failed at /cache/build/tester-amdci4-14/julialang/julia-master/julia-57caa57559/share/julia/stdlib/v1.12/Pkg/test/new.jl:820
  Expression: ((Pkg.dependencies())[exuuid]).version == v"0.5.3"
   Evaluated: v"0.5.5" == v"0.5.3"
Error in testset Pkg:
Test Failed at /cache/build/tester-amdci4-14/julialang/julia-master/julia-57caa57559/share/julia/stdlib/v1.12/Pkg/test/pkg.jl:280
  Expression: ((Pkg.dependencies())[TEST_PKG.uuid]).version > old_v
   Evaluated: v"0.5.5" > v"0.5.5"
Error in testset Pkg:
Test Failed at /cache/build/tester-amdci4-14/julialang/julia-master/julia-57caa57559/share/julia/stdlib/v1.12/Pkg/test/repl.jl:88
  Expression: ((Pkg.dependencies())[TEST_PKG.uuid]).version > v
   Evaluated: v"0.5.5" > v"0.5.5"
```

IanButterworth commented:
That's fixed on master

IanButterworth committed:
Merge branch 'master' into kp/add-trace-dispatch

IanButterworth commented:
Just sharing a thought that if the user wants both `--trace-compile` and `--trace-dispatch` and to be able to tell which is which they can send them to differently named files. 
If they wanted to use stderr they could do `--trace-compile=stderr --trace-compile-timing --trace-dispatch=stderr` and the compiled ones would show with the timing comment first.

Maybe that's overthinking it but at least it's possible if desired.

IanButterworth commented:
Merge? @JeffBezanson

vtjnash commented on src/gf.c:
```diff
@@ -3466,6 +3510,15 @@ jl_value_t *jl_gf_invoke_by_method(jl_method_t *method, jl_value_t *gf, jl_value
+        uint8_t miflags = jl_atomic_load_relaxed(&mfunc->flags);
+        uint8_t was_dispatched = miflags & JL_MI_FLAGS_MASK_DISPATCHED;
+        if (!was_dispatched) {
+            miflags |= JL_MI_FLAGS_MASK_DISPATCHED;
+            jl_atomic_store_relaxed(&mfunc->flags, miflags);
```
Shouldn't these require an `jl_atomic_fetch_or_relaxed` now, since you bit-packed them?
